lib/que/worker.rb

```diff
@@ -5,21 +5,80 @@
 
 module Que
+  class JobTimeoutError < StandardError; end
+
   class Worker
     # Defines the time a worker will wait before checking Postgres for its next job
     DEFAULT_QUEUE = ''
     DEFAULT_WAKE_INTERVAL = 5
     DEFAULT_LOCK_CURSOR_EXPIRY = 0 # seconds
+    DEFAULT_STOP_TIMEOUT = 5 # seconds
+
+    # Start worker_count number of threaded workers, returning a proc that can be called
+    # to stop the workers from working.
+    def self.start_workers(worker_count, metrics_labels: {}, **kwargs)
+      Que.logger.info(msg: 'Starting workers', event: 'que.worker.start', worker_count: worker_count)
+
+      # Create each worker with a distinct ID in order to differentiate each workers
+      # prometheus metrics.
+      workers = Array.new(worker_count) do |id|
+        Worker.new(**kwargs.merge(metrics_labels: metrics_labels.merge(worker: id)))
+      end
+
+      worker_threads = workers.map { |worker| Thread.new { worker.work_loop } }
+
+      -> (stop_timeout = DEFAULT_STOP_TIMEOUT) do
+        Que.logger.info(msg: "Asking workers to finish", event: "que.worker.finish_wait")
+
+        workers.each(&:stop!)
+
+        # Asynchronously stop all workers, sending a join method call with a timeout. This
+        # is done asynchronously to ensure we start the timeout at the same time for each
+        # worker, instead of applying them cumulatively.
+        worker_threads.each_with_index do |thread, idx|
+          Thread.new do
+            unless thread.join(stop_timeout)
+              Que.logger.info(
+                msg: "Worker still running - forcing it to stop",
+                worker: idx, event: "que.worker.finish_timeout",
+              )
+
+              # Caveat that this API (Thread.raise) can be dangerous in Ruby. It can leave
+              # resources in strange states as the exception is trapped immediately in the
+              # target thread, potentially during code that expected no interruptions.
+              #
+              # We can expect this exception to be raised just before our process receives
+              # a SIGKILL, so any hanging resources are probably about to be claimed by
+              # the OS after process death. Given our impending doom, the benefit of
+              # having Que track timeouts against the affected job seems worth the risk of
+              # short-lived undefined behaviour.
+              #
+              # If we begin to see issues with this approach (strange TCP connection
+              # states, connection pool issues, etc) then we should re-evaluate.
+              thread.raise(JobTimeoutError, "Job exceeded timeout when requested to stop")
+            end
+          end
+        end
+
+        # We now know the thread is either finished or has raised an exception. The final
+        # join gives the worker thread time to finish its work loop and record the timeout
+        # exception against the job it was working.
+        worker_threads.each(&:join)
```

Comment:
Do we also need to specify a join timeout again here?

Reply:
I considered this, and didn't come to a strong conclusion. By this time you've stopped the worker, which ensures it won't pickup any new jobs, so it should be finishing the final iteration of its work loop which should be very fast.

If something had gone terribly wrong- like a job catching the exception we raised- then the worker may still be working, but it feels like it's a stretch to defend against it here. I'd like to leave it as is for the moment, given there's little else we can do at this point to handle the timeout other than violently kill the workers, which our orchestrator is about to do.